packages/engine/Source/Scene/Model/PntsLoader.js

```diff
@@ -572,6 +581,465 @@ function makeStructuralMetadata(parsedContent, customAttributeOutput) {
   });
 }
 
+function findBoundingBox(positions) {
+  const numberOfPoints = positions.length / 3;
+  let minx = Infinity;
+  let maxx = -Infinity;
+  let miny = Infinity;
+  let maxy = -Infinity;
+  let minz = Infinity;
+  let maxz = -Infinity;
+  for (let i = 0; i < numberOfPoints; i++) {
+    if (minx > positions[3 * i]) {
+      minx = positions[3 * i];
+    }
+    if (maxx < positions[3 * i]) {
+      maxx = positions[3 * i];
+    }
+    if (miny > positions[3 * i + 1]) {
+      miny = positions[3 * i + 1];
+    }
+    if (maxy < positions[3 * i + 1]) {
+      maxy = positions[3 * i + 1];
+    }
+    if (minz > positions[3 * i + 2]) {
+      minz = positions[3 * i + 2];
+    }
+    if (maxz < positions[3 * i + 2]) {
+      maxz = positions[3 * i + 2];
+    }
+  }
+  const ret = new Float64Array(6);
+  ret[0] = minx;
+  ret[1] = maxx;
+  ret[2] = miny;
+  ret[3] = maxy;
+  ret[4] = minz;
+  ret[5] = maxz;
+  return ret;
+}
+
+function make3DTree(positions) {
+  const numberOfPoints = positions.length / 3;
+  const indexedTree = new Int32Array(numberOfPoints * 2);
+  const boundingBox = findBoundingBox(positions);
+  for (let i = 1; i < numberOfPoints; i++) {
+    const curNode = 0;
+    const depth = 1;
+    appendToTree3D(indexedTree, positions, i, curNode, depth, boundingBox);
+  }
+  return [indexedTree, boundingBox];
+}
+
+function appendToTree3D(
+  indexedTree,
+  positions,
+  pointIndex,
+  curNode,
+  depth,
+  boundingBox
+) {
+  const curx = positions[curNode * 3];
+  const cury = positions[curNode * 3 + 1];
+  const curz = positions[curNode * 3 + 2];
+  const pointx = positions[pointIndex * 3];
+  const pointy = positions[pointIndex * 3 + 1];
+  const pointz = positions[pointIndex * 3 + 2];
+  if (depth % 3 === 0) {
```

Comment:
one way to make this a bit more consistent is assign an enum

```
enum Axis {
x = 0,
y,
z}
```

so instead of depth %3 === 0, you can do:
```
const axis = depth % 3
const searchPoint = positions[pointIndex * 3 + axis]
const curPoint = positions[curNode * 3 +axis]
```
and this will help simplify/ensure that you're keeping your variable straight

Reply:
*this is a party trick but the principle is generally applicable to any language